Added new DoD topics #9282 (#9)
AdobeDocs/commerce-contributor: src/pages/guides/code-contributions/tco-analysis.md

---
title: Definition of Done | TCO Analysis

> description: Total cost of ownership and the Commerce definition of done.

Reviewer: I forgot to rewrite this before submitting my original review. Please rewrite.

---

> ## Overview

Reviewer: ```suggestion
# Total cost of ownership (TCO) analysis
```

All functional changes must be analyzed for the impact on merchant total cost of ownership (TCO) and the results of the analysis should be added to the pull request or issue description to meet the "Definition of Done" (DoD).

The goal of TCO analysis is to identify:

-  If the changes have impact on TCO
-  The value of the impact, that consists of:
   -  The level of the change (patch/minor/major)
   -  Potential number of affected extensions/customizations (several/many/most/all)

> ### Changes that have impact on TCO

Reviewer: ```suggestion
## Changes that have impact TCO
```

Changes can impact total cost of ownership if the extensions or customizations, established processes or integrations within the projects build on Adobe Commerce framework may need to be adjusted to accommodate the changes.

Some changes that have impact on TCO:

-  Result in SVC build failure on pull request
-  JavaScript changes (not covered by SVC build)
   -  Introduced or removed javascript events, classes, methods, properties or parameters
   -  Changed javascript event ordering
-  URL/Request changes
   -  Removed URL paths
   -  Removed or introduced required or optional request parameters
-  CLI commands changes
   -  Introduced or removed/renamed console commands, arguments or return codes
-  Events changes
   -  Removed or added events and their arguments
-  Removing, changing or adding default parameter values for methods
-  Removing or adding traits
-  Changing arguments or return value format of methods
-  Changing topics or consumers in message queue
-  Any layout or block changes
-  Removing or renaming of any files

> ### Level of change

Reviewer: ```suggestion
## Level of change
```

| GraphQL schema                         | Removed type, field, argument, interface, directive, value, location    | MAJOR |
|                                        | Added required argument, input field, directive argument   | MAJOR |
|                                        | Changed kind of argument, type, field| MAJOR |

### Potential number of affected customizations

In order to identify the scope of the impact, usage of the changed code should be detected in the available code/extensions/customizations.

The analysis result should indicate the proportion of reviewed modules to the modules referencing (using) the changed code.

### TCO Analysis result example

The analysis should indicate all the changes that have TCO impact or state that there is "No TCO impact".

For example, if the non-api "SomeClass::someMethod" was changed and the new required argument was added to the CLI command, the analysis result may look like:

```text
Changes have TCO impact.
Change to SomeClass::someMethod is PATCH level but impacts 5 out of 10 reviewed custom modules.
Change to CLI command is MAJOR level, it impacts 2 out of 10 reviewed custom modules.
```

> A custom module is a [Magento module][1] that is not part of Adobe Commerce product.

Reviewer: ```suggestion
A custom module is a [module](https://devdocs.magento.com/architecture/archi_perspectives/components/modules/mod_intro.html) that is not part of the Adobe Commerce product.
```

A change impacts a custom module if its code needs to be changed in response.

> [1]: https://devdocs.magento.com/architecture/archi_perspectives/components/modules/mod_intro.html

Reviewer: ```suggestion
```
